crates/lsp/src/lib.rs

```diff
@@ -167,57 +130,52 @@ fn on_shutdown(
 }
 
 fn on_code_lens_request(
-    state: &mut LspState,
+    _state: &mut LspState,
     params: CodeLensParams,
 ) -> impl Future<Output = Result<Option<Vec<CodeLens>>, ResponseError>> {
-    let file_path = &params.text_document.uri.to_file_path().unwrap();
+    let file_path = params.text_document.uri.to_file_path().unwrap();
+    let program_dir = file_path.parent().unwrap();
 
-    let mut context = match &state.root_path {
-        Some(root_path) => {
-            let fm = FileManager::new(root_path);
-            let graph = CrateGraph::default();
-            Context::new(fm, graph)
-        }
-        None => {
-            let err = ResponseError::new(
-                ErrorCode::REQUEST_FAILED,
-                "Unable to determine the project root path",
-            );
-            return future::ready(Err(err));
-        }
-    };
+    let toml_path = find_package_manifest(program_dir).unwrap();
+    let workspace = resolve_workspace_from_toml(&toml_path, None).unwrap();
 
-    let crate_id = prepare_crate(&mut context, file_path);
+    let mut lenses: Vec<CodeLens> = vec![];
 
-    // We ignore the warnings and errors produced by compilation for producing codelenses
-    // because we can still get the test functions even if compilation fails
-    let _ = check_crate(&mut context, crate_id, false);
+    for package in &workspace {
```

Comment:
Not sure what to do if we encounter a workspace manifest instead of a package manifest.

Reply:
rust-analyzer just ignores a `.rs` file outside of the workspace. I only have access to vscode-specific stuff like snippets, etc. No compiling or syntax errors are available. No warning is made that it is outside of the workspace.

Reply:
I feel simplistic approach of allowing top-level workspace is enough for now.